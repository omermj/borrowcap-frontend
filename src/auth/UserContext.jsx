import React from "react";

/** Context to provide currentUser thoughout the app */

const UserContext = React.createContext();

export default UserContext;
