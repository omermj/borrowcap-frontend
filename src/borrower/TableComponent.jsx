import Table from "react-bootstrap/Table";
import { formatCurrency, formatDate, formatPercent } from "../helpers/format";
import "./TableComponent.css";

const TableComponent = ({ headers, tableData }) => {
  const valueTypes = Object.values(headers);

  return (
    <div>
      <Table className="table" striped hover bordered>
        <thead>
          <tr>
            {Object.keys(headers).map((headerKey, idx) => (
              <th key={idx}>{headers[headerKey].label}</th>
            ))}
          </tr>
        </thead>
        <tbody>
          {tableData.map((row, rowIdx) => (
            <tr key={rowIdx}>
              {Object.keys(headers).map((headerKey, cellIdx) => {
                const formatter = headers[headerKey].formatter;
                const cellValue = row[headerKey];
                return (
                  <td key={cellIdx}>
                    {typeof formatter === "function"
                      ? formatter(cellValue)
                      : cellValue}
                  </td>
                );
              })}
            </tr>
          ))}
        </tbody>
      </Table>
    </div>
  );
};

export default TableComponent;
