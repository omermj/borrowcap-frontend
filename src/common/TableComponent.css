/* th, td {
  vertical-align: middle;
  color: red;
} */

.table td.fit,
.table th.fit {
  white-space: nowrap;
  width: 1%;
}

.table th {
  background-color: rgb(223, 223, 223);
  /* color: white; */

}