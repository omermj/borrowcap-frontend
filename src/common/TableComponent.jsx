import { Table, Button } from "react-bootstrap";
import { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import "./TableComponent.css";

const TableComponent = ({ headers, tableData }) => {
  const [data, setData] = useState(tableData);

  const TableButton = ({ link, label, onClick, icon, testId }) => (
    <Link
      data-testid={testId ? testId : null}
      className="btn btn-dark btn-sm"
      to={onClick ? null : link}
      size="sm"
      onClick={onClick ? onClick : null}
    >
      {icon ? icon() : label}
    </Link>
  );

  useEffect(() => {
    if (tableData) {
      setData(tableData);
    }
  }, [tableData]);

  return (
    <div className="table-responsive" >
      <Table className="table" striped hover borderless>
        <thead>
          <tr className="align-middle">
            {Object.keys(headers).map((headerKey, idx) => (
              <th key={idx}>{headers[headerKey].label}</th>
            ))}
          </tr>
        </thead>
        <tbody>
          {data.map((row, rowIdx) => (
            <tr key={rowIdx} className="align-middle">
              {Object.keys(headers).map((headerKey, cellIdx) => {
                const formatter = headers[headerKey].formatter;
                const cellValue = row[headerKey];
                return (
                  <td key={cellIdx}>
                    {formatter === "button" ? (
                      <TableButton
                        testId={headers[headerKey].testId}
                        label={
                          headers[headerKey].icon
                            ? headers[headerKey].icon()
                            : headers[headerKey].label
                        }
                        link={`${headers[headerKey].link}/${row.id}`}
                        onClick={
                          headers[headerKey].onClick
                            ? async () =>
                                await headers[headerKey].onClick(row.id)
                            : null
                        }
                      />
                    ) : typeof formatter === "function" ? (
                      formatter(cellValue)
                    ) : (
                      cellValue
                    )}
                  </td>
                );
              })}
            </tr>
          ))}
        </tbody>
      </Table>
    </div>
  );
};

export default TableComponent;
