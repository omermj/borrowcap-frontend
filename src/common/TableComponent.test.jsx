import { render } from "@testing-library/react";
import TableComponent from "./TableComponent";
import { BrowserRouter } from "react-router-dom";
import UserContext from "../auth/UserContext";

// smoke test
it("renders without crashing", () => {
  const headers = {
    id: { label: "ID", formatter: "none" },
    amount: { label: "Amount", formatter: "none" },
  };
  const data = [1, 10000];

  render(
    <BrowserRouter>
      <UserContext.Provider value={{ currentUser: { username: "test" } }}>
        <TableComponent headers={headers} tableData={data} />
      </UserContext.Provider>
    </BrowserRouter>
  );
});

// snapshot test
it("matches snapshot", () => {
  const headers = {
    id: { label: "ID", formatter: "none" },
    amount: { label: "Amount", formatter: "none" },
  };
  const data = [1, 10000];
  const { asFragment } = render(
    <BrowserRouter>
      <UserContext.Provider value={{ currentUser: { username: "test" } }}>
        <TableComponent headers={headers} tableData={data} />
      </UserContext.Provider>
    </BrowserRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
