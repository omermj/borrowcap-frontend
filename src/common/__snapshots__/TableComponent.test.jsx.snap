// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <table
      class="table table table-striped table-borderless table-hover"
    >
      <thead>
        <tr
          class="align-middle"
        >
          <th>
            ID
          </th>
          <th>
            Amount
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="align-middle"
        >
          <td />
          <td />
        </tr>
        <tr
          class="align-middle"
        >
          <td />
          <td />
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
